source/base/parameter_handler.cc

```diff
@@ -737,6 +737,38 @@ namespace
           }
       }
   }
+
+  void
+  recursively_keep_non_default(const boost::property_tree::ptree &tree_in,
+                               boost::property_tree::ptree       &tree_out)
+  {
+    for (const auto &p : tree_in)
+      {
+        if (is_parameter_node(p.second))
+          {
+            const std::string value = p.second.get<std::string>("value");
+
+            if (value != p.second.get<std::string>("default_value"))
+              tree_out.put_child(p.first, p.second);
+          }
+        else if (is_alias_node(p.second))
+          {
+            // nothing to do
+          }
+        else
+          {
+            boost::property_tree::ptree temp;
+
+            if (const auto val = p.second.get_value_optional<std::string>())
+              temp.put_value<std::string>(*val);
+
+            recursively_keep_non_default(p.second, temp);
```

Comment:
I don't understand this part of the code completely. `tree_out` is a pure output argument of this function. So it does not matter that you *first* copy `val`, and *then* call the function recursively, right? It could be the other way around? I think this might be easier to read if you added comments to before the `if` statement of the sort "First, copy the top node of this tree, if it exists" and below "And then also copy the changed nodes of the remainder of the tree recursively" (assuming I understood what the code does correctly).